library/choose_disk.py

```diff
@@ -0,0 +1,453 @@
+#!/usr/bin/python
+
+from ansible.module_utils.basic import AnsibleModule
+import logging
+import os
+import re
+import subprocess
+
+DOCUMENTATION = '''
+---
+module: choose_disk
+author: Erwan Velu <erwan@redhat.com>
+short_description: Choose disks based on their features
+description:
+    To be completed
+'''
+
+
+def _equal(left, right):
+    return left == right
+
+
+def _gt(left, right):
+    return float(left) > float(right)
+
+
+def _lt(left, right):
+    return float(left) < float(right)
+
+
+_REGEXP = re.compile(r'^([^(]+)'          # function name
+                     r'\(\s*([^,]+)'      # first argument
+                     r'(?:\s*,\s*(.+))?'  # remaining optional arguments
+                     r'\)$')              # last parenthesis
+
+
+logger = logging.getLogger('choose_disk')
+
+
+def convert_units(value):
+    ''' Convert units to ease comparaison '''
+    value = str(value).lower().strip()
+    storage_units = {
+            'kb': 1024,
+            'kib': 1000,
+            'mb': 1024*1024,
+            'mib': 1000*1000,
+            'gb': 1024*1024*1024,
+            'gib': 1000*1000*1000,
+            'tb': 1024*1024*1024*1024,
+            'tib': 1000*1000*1000*1000,
+            'pb': 1024*1024*1024*1024*1024,
+            'pib': 1000*1000*1000*1000*1000
+    }
+
+    # Units are storage units
+    for size in storage_units.keys():
+        if value.endswith(size):
+            real_value, unit = value.split(" ")
+            return str(float(real_value) * storage_units[unit])
+
+    return value
+
+
+def get_keys_by_ceph_order(physical_disks):
+    '''
+    Return a list of keys where ceph disks are reported first
+    while keeping the list sorted
+    '''
+    ceph_disks = []
+    non_ceph_disks = []
+    for physical_disk in sorted(physical_disks):
+        if "ceph_prepared" in physical_disks[physical_disk]:
+            ceph_disks.append(physical_disk)
+        else:
+            non_ceph_disks.append(physical_disk)
+
+    return ceph_disks + non_ceph_disks
+
+
+def find_match(physical_disks, lookup_disks, module=None):
+    ''' Find a set of matching devices in physical_disks
+    '''
+    matched_devices = {}
+    ignored_devices = []
+
+    logger.info("Looking for matches")
+    # Inspecting every disk we search for
+    for disk in sorted(lookup_disks):
+
+        current_lookup = dict(lookup_disks[disk])
+        infinite = False
+        current_type = ""
+        if "infinite" in current_lookup:
+            infinite = True
+            del current_lookup["infinite"]
+
+        # We cannot keep the disk type as a feature to lookup
+        if "ceph_type" in current_lookup.keys():
+            current_type = current_lookup["ceph_type"]
+            del current_lookup["ceph_type"]
+
+        if len(ignored_devices) == len(physical_disks):
+            logger.info(" Skipping %s as no more free devices to match" % (disk))
```

Comment:
When using the logging module directly it's safer to use it's built-in method for string interpolation instead of the ``"%s" % var`` approach.  This is safer because if you use a variable there that does not interpolate correctly the code will throw an exception and halt execution. If you use the logging modules built-in way then it will catch that exception and report it as part of the logging message without stopping code execution.

That method would look like:

```
logger.info(" Skipping %s as no more free devices to match", disk)
```

Here's an example the exposes the problem I mentioned:

```
import logging

logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)
logger = logging.getLogger()

logger.info('normal logging line')
logger.info('normal logging line, with some correct args: %s %s', 'arg1', 'arg2')
logger.info('normal logging line, with some incorrect args: %s %s %s %s', 'arg1', 'arg2')
logger.info('execution does not halt with the previous invalid line')
logger.info('invalid formatting without logging: %s %s' % 'arg1')
logger.info('this line will never show because the previous line is broken')
```

Could you please all the logging lines to use the approach I mentioned?